File: common/src/main/java/cc/turtl/chiselmon/system/dashboard/api/endpoints/SpeciesHandler.java
package cc.turtl.chiselmon.system.dashboard.api.endpoints;

import cc.turtl.chiselmon.system.dashboard.api.ApiHandler;
import cc.turtl.chiselmon.system.tracker.EncounterDatabase;
import com.sun.net.httpserver.HttpExchange;

import java.io.IOException;
import java.util.List;

public class SpeciesHandler extends ApiHandler {

    private static final int DEFAULT_LIMIT = 20;

    private record SpeciesEntry(String species, long count) {
    }

    private record SpeciesResponse(List<SpeciesEntry> species) {
    }

    public SpeciesHandler(EncounterDatabase db) {
        super(db);
    }

    @Override
    public void handle(HttpExchange exchange) throws IOException {
        handleRequest(exchange, (timeRange, params) -> {
            int limit = parseIntParam(params, "limit", DEFAULT_LIMIT);

            List<SpeciesEntry> species = query("encounters")
                    .timeRange(timeRange)
                    .select("species, COUNT(*) as cnt")
                    .groupBy("species")
                    .orderBy("cnt DESC")
                    .limit(limit)
                    .fetchList(rs -> new SpeciesEntry(rs.getString("species"), rs.getLong("cnt")));

            return new SpeciesResponse(species);
        });
    }
}

File: common/src/main/kotlin/cc/turtl/chiselmon/client/system/dashboard/api/endpoints/SpeciesHandler.kt
package cc.turtl.chiselmon.client.system.dashboard.api.endpoints

import cc.turtl.chiselmon.client.system.dashboard.api.ApiHandler
import cc.turtl.chiselmon.client.system.tracker.EncounterDatabase
import com.sun.net.httpserver.HttpExchange

class SpeciesHandler(db: EncounterDatabase) : ApiHandler(db) {

    private data class SpeciesEntry(val species: String, val count: Long)
    private data class SpeciesResponse(val species: List<SpeciesEntry>)

    override fun handle(exchange: HttpExchange) {
        handleRequest(exchange) { timeRange, params ->
            val limit = parseIntParam(params, "limit", DEFAULT_LIMIT)

            val species = query("encounters")
                .timeRange(timeRange)
                .select("species, COUNT(*) as cnt")
                .groupBy("species")
                .orderBy("cnt DESC")
                .limit(limit)
                .fetchList { rs -> SpeciesEntry(rs.getString("species"), rs.getLong("cnt")) }

            SpeciesResponse(species)
        }
    }

    companion object {
        private const val DEFAULT_LIMIT = 20
    }
}
